feg/gateway/services/n7_n40_proxy/n7/n7.go

```diff
@@ -39,3 +50,509 @@ func NewN7Client(cfg *PCFConfig) (*n7_sbi.ClientWithResponses, error) {
+}
+
+func GetPolicyRuleProto(
+	policyDecision *n7_sbi.SmPolicyDecision,
+	pccRule *n7_sbi.PccRule,
+) *protos.PolicyRule {
+	policyRule := &protos.PolicyRule{
+		Id: pccRule.PccRuleId,
+	}
+	if pccRule.Precedence != nil {
+		policyRule.Priority = uint32(*pccRule.Precedence)
+	}
+	// For all Ref*Data Maximum one entry is expected.
+	// These are declared as arrays for future compatibility
+	monKeyPresent := false
+	ratingGrpPresent := false
+	dynamicRule := false
+	if pccRule.RefUmData != nil && len(*pccRule.RefUmData) > 0 && policyDecision.UmDecs != nil {
+		umData, found := policyDecision.UmDecs.Get((*pccRule.RefUmData)[0])
+		if found {
+			policyRule.MonitoringKey = []byte(umData.UmId)
+			monKeyPresent = true
+			dynamicRule = true
+		}
+	}
+	if pccRule.RefChgData != nil && len(*pccRule.RefChgData) > 0 && policyDecision.ChgDecs != nil {
+		chgData, found := policyDecision.ChgDecs.Get((*pccRule.RefChgData)[0])
+		if found {
+			policyRule.RatingGroup = uint32(*chgData.RatingGroup)
+			policyRule.ServiceIdentifier = &protos.ServiceIdentifier{
+				Value: uint32(*chgData.ServiceId),
+			}
+			policyRule.Offline = getSbiBool(chgData.Offline)
+			policyRule.Online = getSbiBool(chgData.Online)
+			ratingGrpPresent = true
+			dynamicRule = true
+		}
+	}
+	policyRule.TrackingType = GetTrackingTypeProto(monKeyPresent, ratingGrpPresent)
+	if pccRule.RefTcData != nil && len(*pccRule.RefTcData) > 0 && policyDecision.TraffContDecs != nil {
+		tcData, found := policyDecision.TraffContDecs.Get((*pccRule.RefTcData)[0])
+		if found {
+			policyRule.Redirect = GetRedirectInfoProto(tcData.RedirectInfo)
+			dynamicRule = true
+		}
+	}
+	if pccRule.RefQosData != nil && len(*pccRule.RefQosData) > 0 && policyDecision.QosDecs != nil {
+		qosData, found := policyDecision.QosDecs.Get((*pccRule.RefQosData)[0])
+		if found {
+			policyRule.Qos = GetFlowQosProto(&qosData)
+			dynamicRule = true
+		}
+	}
+	if pccRule.FlowInfos != nil {
+		for _, flowInfo := range *pccRule.FlowInfos {
+			if flowInfo.FlowDescription != nil {
+				flow, err := policydb.GetFlowDescriptionFromFlowString(string(*flowInfo.FlowDescription))
+				if err != nil {
+					glog.Errorf("Could not get flow for description %s : %s", *flowInfo.FlowDescription, err)
+				} else {
+					policyRule.FlowList = append(policyRule.FlowList, flow)
+				}
+			}
+		}
+	}
+	if !dynamicRule {
+		// not a dynamic rule
+		return nil
+	}
+	return policyRule
+}
+
+func GetTrackingTypeProto(monKeyPresent bool, ratingGrpPresent bool) protos.PolicyRule_TrackingType {
+	if monKeyPresent && ratingGrpPresent {
+		return protos.PolicyRule_OCS_AND_PCRF
+	} else if monKeyPresent && !ratingGrpPresent {
+		return protos.PolicyRule_ONLY_PCRF
+	} else if !monKeyPresent && ratingGrpPresent {
+		return protos.PolicyRule_ONLY_OCS
+	} else {
+		return protos.PolicyRule_NO_TRACKING
+	}
+}
+
+func GetRedirectInfoProto(redirInfo *n7_sbi.RedirectInformation) *protos.RedirectInformation {
```

Comment:
apply here and in the rest of the functions pointers receiver either to tproto or sbi interface

Reply:
See other responses related to this.